tests/functions.sh

```diff
@@ -243,12 +360,326 @@ function delete_nic {
+        security_group: \"$sg\"
+      }
+    }
+  }
+}
+messages {
+  revision: 0
+  message_0 {
+    request {
+      sg_add {
+        id: \"sg-1\"
+        rule {
+          direction: INBOUND
+          protocol: -1
+          cidr {
+            address: \"0.0.0.0\"
+            mask_size: 0
+          }
+        }
+      }
+    }
+  }
+}
+" > $f
+    request $but_id $nic_id $f
+
+    if ! test -e /tmp/qemu-vhost-nic-$nic_id ; then
+        echo "client failed: we should have a socket in /tmp/qemu-vhost-nic-$nic_id"
+        clean_all
+        exit 1
+    fi
+}
+
+function update_sg_rule_port_open {
+    but_id=$1
+    nic_id=$2
+    protocol=$3
+    port=$4
+    sg=$5
+    f=/tmp/butterfly-client.req
+    echo "update nic $nic_id with only port $port opened in butterfly $but_id in vni $vni"
+
+    echo -e "messages {
+  revision: 0
+  message_0 {
+    request {
+      nic_update {
+        id: \"nic-$nic_id\"
+        ip: \"42.0.0.$nic_id\"
+        ip_anti_spoof: true
+        security_group: \"$sg\"
+      }
+    }
+  }
+}
+messages {
+  revision: 0
+  message_0 {
+    request {
+      sg_add {
+        id: \"$sg\"
+        rule {
+          direction: INBOUND
+          protocol: $protocol
+          port_start: $port
+          port_end: $port
+          cidr {
+            address: \"0.0.0.0\"
+            mask_size: 0
+          }
+        }
+      }
+    }
+  }
+}
+" > $f
+    request $but_id $nic_id $f
+
+    if ! test -e /tmp/qemu-vhost-nic-$nic_id ; then
+        echo "client failed: we should have a socket in /tmp/qemu-vhost-nic-$nic_id"
+        clean_all
+        exit 1
+    fi
+}
+
+function update_nic_sg {
+    but_id=$1
+    nic_id=$2
+    sg=$3
+    echo "Update SG of nic $nic_id in butterfly $1"  
+    f=/tmp/butterfly-client.req
+
+    echo -e "messages {
+  revision: 0
+  message_0 {
+    request {
+      nic_update {
+        id: \"nic-$nic_id\"
+        ip: \"42.0.0.$nic_id\"
+        ip_anti_spoof: true
+        security_group: \"$sg\"
+      }
+    }
+  }
+}
+" > $f
+    request $but_id $nic_id $f
+
+    if ! test -e /tmp/qemu-vhost-nic-$nic_id ; then
+        echo "client failed: we should have a socket in /tmp/qemu-vhost-nic-$nic_id"
+        clean_all
+        exit 1
+    fi
+}
+
+function delete_nic_sg {
+    but_id=$1
+    nic_id=$2
+    sg=$3
+    f=/tmp/butterfly-client.req
+    echo "delete SG of nic $nic_id in butterfly $but_id"
+
+    echo -e "messages {
+  revision: 0
+  message_0 {
+    request {
+      sg_del: \"$sg\"
+    }
+  }
+}
+" > $f
+    request $but_id $nic_id $f
+}
+
+function delete_rule_full_open {
+    but_id=$1
+    nic_id=$2
+    sg=$3
+    echo "delete nic $nic_id rules full open in butterfly $but_id"
+    f=/tmp/butterfly-client.req
+
+    echo -e "messages {
+  revision: 0
+  message_0 {
+    request {
+      nic_update {
+        id: \"nic-$nic_id\"
+        ip: \"42.0.0.$nic_id\"
+        ip_anti_spoof: true
+        security_group: \"$sg\"
+      }
+    }
+  }
+}
```

Comment:
same here: Why do you update the nic ? If a NIC is already configured to use a SG, you don't need to nic update